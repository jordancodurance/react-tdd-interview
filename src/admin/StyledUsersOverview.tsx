import {styled} from "@mui/material/styles";
import TableCell, {tableCellClasses} from "@mui/material/TableCell";
import TableRow from "@mui/material/TableRow";
import TableContainer from "@mui/material/TableContainer";
import Paper from "@mui/material/Paper";
import Table from "@mui/material/Table";
import TableHead from "@mui/material/TableHead";
import TableBody from "@mui/material/TableBody";
import React, {ReactNode} from "react";

const StyledTableCell = styled(TableCell)(({theme}) => ({
    [`&.${tableCellClasses.head}`]: {
        backgroundColor: '#FFCCBC',
        color: '#C62828',
        border: '1px solid black',
        fontWeight: 'bold',
        fontSize: '1em'
    },
    [`&.${tableCellClasses.body}`]: {
        fontSize: 14,
        border: '1px solid black'
    },
}));

const StyledTableRow = styled(TableRow)(({theme}) => ({
    '&:nth-of-type(even)': {
        backgroundColor: theme.palette.action.hover,
    },
}));

export const OverviewTable = ({children}: { children: ReactNode }) => (
    <TableContainer component={Paper}>
        <Table>
            {children}
        </Table>
    </TableContainer>
);

export const Header = ({children}: { children: ReactNode }) => (
    <TableHead>
        <TableRow>
            {children}
        </TableRow>
    </TableHead>
);

export const Body = ({children}: { children: ReactNode }) => (
    <TableBody>
        {children}
    </TableBody>
);

export const Row = ({children}: { children: ReactNode }) => (
    <StyledTableRow>{children}</StyledTableRow>
);

export const Cell = ({children}: { children: ReactNode }) => (
    <StyledTableCell>{children}</StyledTableCell>
);

export const CenteredCell = ({children}: { children: ReactNode }) => (
    <StyledTableCell align="center">{children}</StyledTableCell>
);

