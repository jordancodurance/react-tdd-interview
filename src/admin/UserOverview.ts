// @ts-ignore
import copy from "./user-overview-copy.json" assert { type: 'json' };

export interface UserOverview {
    name: string;
    rating: number;
    country: string;
    submittedOn: string;
    verificationStatus: VERIFICATION_STATUS;
}

export const VERIFICATION_STATUS = {
    VERIFIED: copy.verifiedStatus,
    UNVERIFIED: copy.unverifiedStatus
} as const;

export type VERIFICATION_STATUS = typeof VERIFICATION_STATUS[keyof typeof VERIFICATION_STATUS];