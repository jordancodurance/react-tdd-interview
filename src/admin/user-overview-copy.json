{
  "nameHeader": "Name",
  "ratingHeader": "Rating",
  "countryHeader": "Country",
  "submissionDate": "Submission Date",
  "verificationHeader": "Verification",
  "verifiedStatus": "Verified",
  "unverifiedStatus": "Unverified"
}