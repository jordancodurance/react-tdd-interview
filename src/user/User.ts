export interface User {
    id: string;
    firstName: string;
    lastName: string;
    submissionDate: string;
    country: string;
    verified: boolean;
    rating: number;
}