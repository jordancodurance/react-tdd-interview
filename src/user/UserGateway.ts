import {User} from "./User";

export interface UserGateway {
    getAll(): Promise<User[]>
}