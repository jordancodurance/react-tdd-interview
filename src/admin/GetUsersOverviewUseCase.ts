import {User} from "./User";
import {UserGateway} from "./UserGateway";
import {UserOverview, VERIFICATION_STATUS} from "./UserOverview";
import {DateConverter} from "../shared/time/DateConverter";

export class GetUsersOverviewUseCase {
    private gateway: UserGateway;

    constructor(gateway: UserGateway) {
        this.gateway = gateway;
    }

    public async execute(): Promise<UserOverview[]> {
        const users = await this.gateway.getAll();
        return users.map(u => this.createUserOverview(u));
    }

    private createUserOverview(user: User): UserOverview {
        return {
            name: `${user.firstName} ${user.lastName}`,
            rating: user.rating,
            country: user.country,
            submittedOn: DateConverter.toDate(user.submissionDate),
            verificationStatus: user.verified ? VERIFICATION_STATUS.VERIFIED : VERIFICATION_STATUS.UNVERIFIED
        };
    }
}