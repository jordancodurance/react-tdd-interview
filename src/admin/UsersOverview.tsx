import React, {useEffect, useState} from 'react';
import {Body, Cell, CenteredCell, Header, OverviewTable, Row} from "./StyledUsersOverview";
import copy from "./user-overview-copy.json";
import {GetUsersOverviewUseCase} from "./GetUsersOverviewUseCase";
import {UserOverview} from "./UserOverview";

const UsersOverview = ({useCase}: { useCase: GetUsersOverviewUseCase }) => {
    const [users, setUsers] = useState<UserOverview[]>([]);

    useEffect(() => {
        useCase.execute().then(u => setUsers(u));
    });

    return (
        <OverviewTable>
            <Header>
                <Cell>{copy.nameHeader}</Cell>
                <CenteredCell>{copy.ratingHeader}</CenteredCell>
                <Cell>{copy.countryHeader}</Cell>
                <Cell>{copy.submissionDate}</Cell>
                <Cell>{copy.verificationHeader}</Cell>
            </Header>
            <Body>
                {users.map((u) => (
                    <Row key={u.name}>
                        <Cell>{u.name}</Cell>
                        <CenteredCell>{u.rating}</CenteredCell>
                        <Cell>{u.country}</Cell>
                        <Cell>{u.submittedOn}</Cell>
                        <Cell>{u.verificationStatus}</Cell>
                    </Row>
                ))}
            </Body>
        </OverviewTable>
    );
};

export default UsersOverview;