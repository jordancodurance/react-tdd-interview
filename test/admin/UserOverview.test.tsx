import {GetUsersOverviewUseCase} from "../../src/admin/GetUsersOverviewUseCase";
import {LocalUserGateway} from "../../src/user/LocalUserGateway";
import renderer from 'react-test-renderer';
import UsersOverview from "../../src/admin/UsersOverview";
import {waitFor} from "@testing-library/react"

describe('User overview should', () => {
    test('retain previous accepted snapshot', async () => {
        const gateway = new LocalUserGateway();
        const useCase = new GetUsersOverviewUseCase(gateway);

        const tree = renderer
            .create(<UsersOverview useCase={useCase}/>)
            .toJSON();

        await waitFor(() => expect(tree).toMatchSnapshot());
    });
});