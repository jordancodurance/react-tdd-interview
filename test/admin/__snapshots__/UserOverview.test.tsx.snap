// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User overview should retain previous accepted snapshot 1`] = `
[
  <style
    dangerouslySetInnerHTML={
      {
        "__html": ".css-rorn0c-MuiTableContainer-root{width:100%;overflow-x:auto;}",
      }
    }
    data-emotion="css rorn0c-MuiTableContainer-root"
  />,
  <style
    dangerouslySetInnerHTML={
      {
        "__html": ".css-11xur9t-MuiPaper-root-MuiTableContainer-root{background-color:#fff;color:rgba(0, 0, 0, 0.87);-webkit-transition:box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;transition:box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;border-radius:4px;box-shadow:0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);width:100%;overflow-x:auto;}",
      }
    }
    data-emotion="css 11xur9t-MuiPaper-root-MuiTableContainer-root"
  />,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
  >
    <style
      dangerouslySetInnerHTML={
        {
          "__html": ".css-rqglhn-MuiTable-root{display:table;width:100%;border-collapse:collapse;border-spacing:0;}.css-rqglhn-MuiTable-root caption{font-family:"Roboto","Helvetica","Arial",sans-serif;font-weight:400;font-size:0.875rem;line-height:1.43;letter-spacing:0.01071em;padding:16px;color:rgba(0, 0, 0, 0.6);text-align:left;caption-side:bottom;}",
        }
      }
      data-emotion="css rqglhn-MuiTable-root"
    />
    <table
      className="MuiTable-root css-rqglhn-MuiTable-root"
      role={null}
    >
      <style
        dangerouslySetInnerHTML={
          {
            "__html": ".css-15wwp11-MuiTableHead-root{display:table-header-group;}",
          }
        }
        data-emotion="css 15wwp11-MuiTableHead-root"
      />
      <thead
        className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        role={null}
      >
        <style
          dangerouslySetInnerHTML={
            {
              "__html": ".css-1q1u3t4-MuiTableRow-root{color:inherit;display:table-row;vertical-align:middle;outline:0;}.css-1q1u3t4-MuiTableRow-root.MuiTableRow-hover:hover{background-color:rgba(0, 0, 0, 0.04);}.css-1q1u3t4-MuiTableRow-root.Mui-selected{background-color:rgba(25, 118, 210, 0.08);}.css-1q1u3t4-MuiTableRow-root.Mui-selected:hover{background-color:rgba(25, 118, 210, 0.12);}",
            }
          }
          data-emotion="css 1q1u3t4-MuiTableRow-root"
        />
        <tr
          className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <style
            dangerouslySetInnerHTML={
              {
                "__html": ".css-1tulurv.MuiTableCell-head{background-color:#FFCCBC;color:#C62828;border:1px solid black;font-weight:bold;font-size:1em;}.css-1tulurv.MuiTableCell-body{font-size:14px;border:1px solid black;}",
              }
            }
            data-emotion="css 1tulurv"
          />
          <style
            dangerouslySetInnerHTML={
              {
                "__html": ".css-yfxmun-MuiTableCell-root{font-family:"Roboto","Helvetica","Arial",sans-serif;font-weight:500;font-size:0.875rem;line-height:1.5rem;letter-spacing:0.01071em;display:table-cell;vertical-align:inherit;border-bottom:1px solid rgba(224, 224, 224, 1);text-align:left;padding:16px;color:rgba(0, 0, 0, 0.87);}.css-yfxmun-MuiTableCell-root.MuiTableCell-head{background-color:#FFCCBC;color:#C62828;border:1px solid black;font-weight:bold;font-size:1em;}.css-yfxmun-MuiTableCell-root.MuiTableCell-body{font-size:14px;border:1px solid black;}",
              }
            }
            data-emotion="css yfxmun-MuiTableCell-root"
          />
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-yfxmun-MuiTableCell-root"
            scope="col"
          >
            Name
          </th>
          <style
            dangerouslySetInnerHTML={
              {
                "__html": ".css-1yfhacw-MuiTableCell-root{font-family:"Roboto","Helvetica","Arial",sans-serif;font-weight:500;font-size:0.875rem;line-height:1.5rem;letter-spacing:0.01071em;display:table-cell;vertical-align:inherit;border-bottom:1px solid rgba(224, 224, 224, 1);text-align:center;padding:16px;color:rgba(0, 0, 0, 0.87);}.css-1yfhacw-MuiTableCell-root.MuiTableCell-head{background-color:#FFCCBC;color:#C62828;border:1px solid black;font-weight:bold;font-size:1em;}.css-1yfhacw-MuiTableCell-root.MuiTableCell-body{font-size:14px;border:1px solid black;}",
              }
            }
            data-emotion="css 1yfhacw-MuiTableCell-root"
          />
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yfhacw-MuiTableCell-root"
            scope="col"
          >
            Rating
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-yfxmun-MuiTableCell-root"
            scope="col"
          >
            Country
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-yfxmun-MuiTableCell-root"
            scope="col"
          >
            Submission Date
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-yfxmun-MuiTableCell-root"
            scope="col"
          >
            Verification
          </th>
        </tr>
      </thead>
      <style
        dangerouslySetInnerHTML={
          {
            "__html": ".css-apqrd9-MuiTableBody-root{display:table-row-group;}",
          }
        }
        data-emotion="css apqrd9-MuiTableBody-root"
      />
      <tbody
        className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        role={null}
      />
    </table>
  </div>,
]
`;
